### PR TITLE
fix(universal-link): more new link format, handle old link format

### LINKED ISSUE #17718: Tapping on PN leads to "Browse" popup on Android
# Bug Report

## Problem 

https://github.com/status-im/status-mobile/assets/4557972/6680f987-4531-4a53-a1fb-d8cd4ee2834d

#### Expected behavior
redirected to chat

#### Actual behavior
"Browse" popup is opened

### Reproduction
Preconditions: `Enable notifications` on onboarding

1. Receive any notification from the contact

### Additional Information

- Status version: nightly 23/10/23
- Operating System: Android

[comment]: # (Please, add logs/notes if necessary)

### LINKED ISSUE #17722: Support new deep link format when push notification
deep link in push notification is `status-im`, need to change them to `status-app`
deep link in push notification is generated in status-go
this PR added some temp fix to ensure e2e test works, need to remove this temp fix once link format in status-go is updated

related code in status-go, there might be more
CleanShot 2023-10-23 at 21 16 28@2x

### LINKED ISSUE #17723: "null is not an object (evaluating 'n.indexOf')" error when pasting link to non-existing profile on add contact page
Might be regression of https://github.com/status-im/status-mobile/pull/17480 cc @yqrashawn 

**Steps:** 

1. Open add contact screen
2. Paste link to non-existing profile, for example https://status.app/u/zQ3shc39i2eXV3UdEQTbnj7vo8yVAn1gKcP8MsLfHjrwKg3Y0
3. Observe the error

**Actual result:** "null is not an object (evaluating 'n.indexOf')" error appears

[logs (11).zip](https://github.com/status-im/status-mobile/files/13071628/logs.11.zip)
photo_2023-10-23 17 14 05

**Expected result:** user not found error appears

### Additional Information

- Status version: nightly 
- Operating System: Android, iOS

[comment]: # (Please, add logs/notes if necessary)

### LINKED ISSUE #17725: Add support for the new profile link format and remove the old one for the "Add a new contact" field
# Bug Report

## Problem 

Follow-up of https://github.com/status-im/status-mobile/pull/17480 @yqrashawn 

We should not support the old user profile format like `https://status.app/u/zQ3shrgSaffLisbiqj2brkDK8Le7wpbpbpzriWd8kPUmdYhs1` when adding a contact using a deep link

#### Expected behavior

Formats that should be supported:

`https://status.app/u/G10A4B0JdgwyRww90WXtnP1oNH1ZLQNM0yX0Ja9YyAMjrqSZIYINOHCbFhrnKRAcPGStPxCMJDSZlGCKzmZrJcimHY8BbcXlORrElv_BbQEegnMDPx1g9C5VVNl0fE4y#zQ3shwQPhRuDJSjVGVBnTjCdgXy5i9WQaeVPdGJD6yTarJQSj`

`https://status.app/u/Ow==#zQ3shwQPhRuDJSjVGVBnTjCdgXy5i9WQaeVPdGJD6yTarJQSj`

#### Actual behavior

https://github.com/status-im/status-mobile/assets/67952253/174fc8d0-7046-4f1f-b043-29981549ca19

### Reproduction

1. Create a profile
2. Open the chat home screen
3. Press the `+` button > `Add a contact`
4. Paste the link

### Additional Information

- Status version: nightly 23/10
- Operating System: Android, iOS

[comment]: # (Please, add logs/notes if necessary)

### PR DESCRIPTION
**Pull request by yqrashawn**
### Summary

- add support for private chat and community join request back
- add support for `/u#<user-chat-key>`
- add support for `/c#<community-id>`
- change old `/c/<user-chat-key>` to new `/c#<user-chat-key>`
- update community-chat notification deep link to new format `/cc/<community-channel-id>#<community-id>`

fixes #17718 
fixes #17725
fixes #17722
fixes #17723

-------

#### Platforms
- Android
- iOS

#### Areas that maybe impacted
anywhere related to old `/u/pubkey` link
anything related to universal/deep link

status: ready

### CODE REVIEW

**Review by ibrkhalil — APPROVED** [review]
Thank you so much!

**Review by BalogunofAfrica — APPROVED** [review]
LGTM

**Comment by pavloburykh** [comment]
@yqrashawn thanx for the PR 

Issue is fixed but only for 1-1 chat. Still reproducing the bug in case of communities and group chats. 


### ISSUE 1 Tapping on community/group chat PN leads to "Browse" popup on Android

**Steps:** 

1. Enable PNs
2. Generate PN in community/group chat
3. tap PN
4. Observe the result

**Actual result:** "Browse" popup is opened


https://github.com/status-im/status-mobile/assets/97245802/9d6a9479-6f65-4550-a0ec-c629539842cc

**Expected result:** user is redirected to community/group chat

**Comment by pavloburykh** [comment]
> ISSUE 1 Tapping on community/group chat PN leads to "Browse" popup on Android

@yqrashawn please ping me up for re-test when you'll have a chance to fix the issue. Thank you.

**Comment by BalogunofAfrica** [comment]
@pavloburykh @yqrashawn is under the weather. Checking it out.

I have enabled for group chats, still trying to figure out why `community` links are opening browse

**Comment by yqrashawn** [comment]
Updates:
created status-go PR to generate correct deep link https://github.com/status-im/status-go/pull/4198
the plan is to support both old and new links for now

although there's a conflict between old and new
- old `cc/` is for community chat
- new `cc/` is for community channel

they have different info in the URL, so I moved the old `cc/` to `ccm/`
tested on android real device and it works

**Comment by pavloburykh** [comment]
> Updates: created status-go PR to generate correct deep link [status-im/status-go#4198](https://github.com/status-im/status-go/pull/4198) the plan is to support both old and new links for now

Thank you @yqrashawn! So once status-go PR is approved we are ready for re-test, correct?

**Comment by yqrashawn** [comment]
@pavloburykh also this one, code changed a lot so I dismissed the approved state of this PR. And I just find out the share profile URL is broken as well. I might need to fix it in this PR as well.

I'll put this PR into draft state and reevaluate what needs to be done.

**Comment by yqrashawn** [comment]
Hi @pavloburykh @qoqobolo  @ibrkhalil @BalogunofAfrica @cammellos 
The scope of this PR has changed. I updated the PR description.
Ready for code review.

File: shadow-cljs.edn
Comment (ibrkhalil) [inline shadow-cljs.edn:114]: You think we should revert these formatting changes?

**Review by ibrkhalil — APPROVED** [review]
lgtm
As always, Great job Rashawn!
Thank you so much for adding backwards compatibility this fast!

File: shadow-cljs.edn
Comment (yqrashawn) [reply shadow-cljs.edn:114]: yes, it's not against the linter

**Comment by yqrashawn** [comment]
No idea why, but seems community fetch is fixed as well cc @cammellos 
Description of the issue in prev PR
> for links that jump to unfetched community, the code call status-go to fetch it from mailserver, but it takes a long time for the data to show up

open link of newly created community


[CleanShot 2023-10-30 at 16.07.23.webm](https://github.com/status-im/status-mobile/assets/15090582/0542371d-158e-4e83-9055-2ac6e6c8619e)

open link of community created days ago

[CleanShot 2023-10-30 at 15.55.41.webm](https://github.com/status-im/status-mobile/assets/15090582/0ced5140-ba48-4b24-9ada-dd6ae95431ec)

**Comment by pavloburykh** [comment]
@yqrashawn thank you!

Small clarification about
> * change old `/c/<user-chat-key>` to new `/c#<user-chat-key>`

did you mean change old `/u/<user-chat-key>` to new `/u#<user-chat-key>` ?

**Comment by pavloburykh** [comment]
@yqrashawn I have pushed a small [commit](https://github.com/status-im/status-mobile/pull/17721/commits/2fb58c82aefb7171ba6b2379317fe8db38b2ad4e) with e2e fixes. Please mind it while rebasing/merging the PR. Thank you. 

Meanwhile I am continuing with testing.

**Comment by pavloburykh** [comment]
@yqrashawn https://github.com/status-im/status-mobile/issues/17723 is still reproducible

### ISSUE 1 "null is not an object (evaluating 'n.indexOf')" error when pasting link to non-existing profile on add contact page

**Steps:** 

1. Open add contact screen
2. Paste valid profile link, for example https://status.app/u#zQ3shprjsCAuNMFp8WWZXHQU2vshutzCbLM6m9aDDarzWsNPU
3. Change key to non existing one, for example https://status.app/u#zQ3shprjsCAuNMFp8WWZXHQU2vshutzCbLM6m9aDDarzWsNPa
4. Observe the error

**Actual result:** "null is not an object (evaluating 'n.indexOf')" error appears


https://github.com/status-im/status-mobile/assets/97245802/2312065e-e4a4-465e-abca-c468437201bd


**Expected result:** user not found error appears

**Comment by yqrashawn** [comment]
> did you mean change old /u/<user-chat-key> to new /u#<user-chat-key> ?

yes @pavloburykh

**Comment by yqrashawn** [comment]
Hi @pavloburykh, ISSUE 1 is fixed.

**Comment by pavloburykh** [comment]
### ISSUE 2 Community is not fetched right after redirection via deeplink

Example of community deeplink https://status.app/c#zQ3shdcrVTC1euFxdBzoyRWCNfZ5jHWdVpEs3DDj69CrSyKYY

**Steps:** 

1. Open the given deeplink
2. See if community is fetched

**Actual result:** community is not fetched. "Community admins will review your request" message is displayed

[logs (12).zip](https://github.com/status-im/status-mobile/files/13206978/logs.12.zip)

photo_2023-10-30 17 32 13

**Expected result:** community home page is fetched.

**Comment by pavloburykh** [comment]
@yqrashawn could you please rebase go branch https://github.com/status-im/status-go/pull/4198 as it is outdated?

**Comment by pavloburykh** [comment]
@yqrashawn what are the usecases of links without ```https```? i.e. deeplinks starting from
```
status-im://
status-app://
status-app://

```

Are we using such links in any place now?

**Comment by yqrashawn** [comment]
Hi @pavloburykh, status-go rebase is done

---

> @yqrashawn what are the usecases of links without https? i.e. deeplinks starting from

`https://status.app` is universal link
`status-app://` is deep link
both android and iOS recommend using universal link over deep link

we are currently using deep links in push notification, basically user click on notification same as open deep link outside app

I also use it in local build to test link handler logic because universal link currently doesn't work for local build. It requires `im.status.ethereum.debug` exists in https://status.app/.well-known/apple-app-site-association

---

for ISSUE2, I created a separate issue for it https://github.com/status-im/status-mobile/issues/17770

**Comment by qoqobolo** [comment]
### ISSUE 3: Links of format `/u/<encoded_data>#<user_chat_key>` don't work when pasting it in the `Add a contact` field

The new format `/u#<user-chat-key>` works fine in this case.
But if you use a real desktop user profile link, it doesn't.


Screenshot 2023-10-31 at 11 08 21
Screenshot 2023-10-31 at 11 08 47

These links also work if you follow them from an external app


https://github.com/status-im/status-mobile/assets/67952253/6cd221ed-1760-4c2d-931e-a4fffc93d0e2

Should we support the `/u/<encoded_data>#<user_chat_key>` format when searching a user via the `Add a contact` field (according to the PR description https://github.com/status-im/status-mobile/issues/17725 is in the scope)?

**Steps:**

1. Copy a desktop user profile link, e.g. https://status.app/u/CwSACgcKBVBhdmxvAw==#zQ3shsKnV5HJMWJR61c6dssWzHszdbLfBoMF1gcLtSQAYdw2d
2. Open Status mobile
3. Go to the `messages` home screen
4. Press the `+` button > `Add a contact`
5. Paste the link

**Expected result:** the user is found
**Actual result:** `Invalid ENS or Chat key`

**Comment by yqrashawn** [comment]
Thanks @qoqobolo 
ISSUE3 is fixed, ready for test again

**Comment by pavloburykh** [comment]
Hey @yqrashawn! Have you pushed some changes in last commit or it was just rebasing?

**Comment by yqrashawn** [comment]
@pavloburykh it's just rebase

**Comment by pavloburykh** [comment]
@yqrashawn could you please clarify which of those links https://yqrashawn.com/_test/status-universal-link.html should be supported now and which should not?

**Comment by yqrashawn** [comment]
- Ignore dev helper
- ethereum one should work
- any status-app:// should work
- any https://status.app should work
- join.status.im should not work
- status-im:// should not work

@pavloburykh

**Comment by pavloburykh** [comment]
In this PR I cannot verify if community/community channels links work correctly because of [ISSUE 2](https://github.com/status-im/status-mobile/pull/17721#issuecomment-1785482410). Most of the time community page is not fetched.

**Comment by pavloburykh** [comment]
> any status-app:// should work

@yqrashawn how should I check this type of deeplinks? 

Let's take status-app://status.app/u#zQ3shwQPhRuDJSjVGVBnTjCdgXy5i9WQaeVPdGJD6yTarJQSj as example. This is deeplink of user's profile. 

**Here what I got when opening link from Safari browser on IOS:** 

User is redirected to the App but Profile does not open. 

https://github.com/status-im/status-mobile/assets/97245802/b0b79200-a321-47db-9d5a-7ac9f77437c5

**In case of Android deeplink cannot be opened via browser**

https://github.com/status-im/status-mobile/assets/97245802/f7a8ea1a-4374-4047-a9a9-7c2fe1dc63ad

**Comment by pavloburykh** [comment]
@yqrashawn I do not want to block PR with this bug. So if you consider it to be not trivial (or not PR related) we can fix it as followup. 

### ISSUE 4 Sometimes PN does not redirect user into correct chat if some other chat was opened at the moment of redirection (Android)

Reproducing on Android 12, Samsing Galaxy A52. Bug is not reproducible every time

**Preconditions:** User A (Android) turns on push notifications

**Steps:** 

1. User A opens 1-1 chat with User B
2. User A closes the app
3. User B sends mention/reply to User A in community channel
4. User A taps on PN 
5. See if User A is redirected into community cahnnel

**Actual result:** sometimes User A remains on 1-1 chat screen and is redirected to the channel. At the same time, the message in channel appears as read despite User A in fact has not opened the channel. 


https://github.com/status-im/status-mobile/assets/97245802/5ecbbb36-928a-4491-beb9-7dd446452a13

[Status-debug-logs (76).zip](https://github.com/status-im/status-mobile/files/13226198/Status-debug-logs.76.zip)


**Expected result:** User A should be redirected to the right chat upon PN.

**Comment by yqrashawn** [comment]
Hi @pavloburykh 
reply to this one https://github.com/status-im/status-mobile/pull/17721#issuecomment-1788735399
`status-app://status.app/u#zQ3shwQPhRuDJSjVGVBnTjCdgXy5i9WQaeVPdGJD6yTarJQSj`
this is a invalid URL, it should be `status-app://u#zQ3shwQPhRuDJSjVGVBnTjCdgXy5i9WQaeVPdGJD6yTarJQSj`
universal and deep link will always trigger app open
when invalid URL opened in app, it should fallback to browse the URL
which ask user about open URL in status or external browser, as showed in the first video

**Comment by pavloburykh** [comment]
> this is a invalid URL, it should be `status-app://u#zQ3shwQPhRuDJSjVGVBnTjCdgXy5i9WQaeVPdGJD6yTarJQSj`

Thank you @yqrashawn Sorry, I was so blind and didn't notice I messed up this link. 

But still, there is a question about Android: can those links be tested somewhere except Status browser? When I open the link in Chrome it just triggers search but not redirection into the app. 

On IOS it works as expected via Safari.

**Comment by pavloburykh** [comment]
### ISSUE 5 Channel links do not redirect user to community homepage

Example of channel link https://status.app/cc/Gx8BAGTimp-rotE3ZfluQAP5iAIf6IgH_7kD1vb-7ZJ2UhQEeSAJBRhzS_UMDzs8WtUEXrEOP-iw_VliFA_yGP1Nhi7ihyEJCb8uy7FKh5Xa0rpmyq83-wcP2CfsB6Ko61HFUm3d_viCWZFnedFqkcNCBDGJIss1vciJnkFTqEAGYcJc9wO3jNDAIZUNOfJRhJlJ6cqI53kBD2GK3KiMG_iyGkS2IEeGLOh6aAquabqCKweGaTqG4zku#zQ3shbmfT3hvh4mKa1v6uAjjyztQEroh8Mfn6Ckegjd7LT3XK

I tried to open link from telegram messenger and 1-1 chat and result is the same:

[logs (13).zip](https://github.com/status-im/status-mobile/files/13226390/logs.13.zip)

https://github.com/status-im/status-mobile/assets/97245802/654afd4c-621c-46cd-afbc-5581fce29cd5

**Comment by pavloburykh** [comment]
@yqrashawn a question regarding community channel link example from your [file](https://yqrashawn.com/_test/status-universal-link.html) 

https://status.app/cc/Ow==#zQ3shYSHp7GoiXaauJMnDcjwU2yNjdzpXLosAWapPS4CFxc11
status-app://cc/Ow==#zQ3shYSHp7GoiXaauJMnDcjwU2yNjdzpXLosAWapPS4CFxc11

How do this links suppose to open specific channel if there is no channel id inside but only community id?

**Comment by yqrashawn** [comment]
> But still, there is a question about Android

just tried to add link to github comment, and it didn't work, github won't make status-app:// link clickable
I just added an input to https://yqrashawn.com/_test/status-universal-link.html
you can use the input and add button to add temp links to the website
links will be gone after refresh
@pavloburykh

**Comment by pavloburykh** [comment]
@yqrashawn could you please clarify were such type of links are generated at this moment https://status.app/u/G10A4B0JdgwyRww90WXtnP1oNH1ZLQNM0yX0Ja9YyAMjrqSZIYINOHCbFhrnKRAcPGStPxCMJDSZlGCKzmZrJcimHY8BbcXlORrElv_BbQEegnMDPx1g9C5VVNl0fE4y#zQ3shwQPhRuDJSjVGVBnTjCdgXy5i9WQaeVPdGJD6yTarJQSj ?

**Comment by yqrashawn** [comment]
> @yqrashawn could you please clarify were such type of links are generated at this moment [status.app/u/G10A4B0JdgwyRww90WXtnP1oNH1ZLQNM0yX0Ja9YyAMjrqSZIYINOHCbFhrnKRAcPGStPxCMJDSZlGCKzmZrJcimHY8BbcXlORrElv_BbQEegnMDPx1g9C5VVNl0fE4y#zQ3shwQPhRuDJSjVGVBnTjCdgXy5i9WQaeVPdGJD6yTarJQSj](https://status.app/u/G10A4B0JdgwyRww90WXtnP1oNH1ZLQNM0yX0Ja9YyAMjrqSZIYINOHCbFhrnKRAcPGStPxCMJDSZlGCKzmZrJcimHY8BbcXlORrElv_BbQEegnMDPx1g9C5VVNl0fE4y#zQ3shwQPhRuDJSjVGVBnTjCdgXy5i9WQaeVPdGJD6yTarJQSj) ?

it's `/u/<encoded_user_data>#<user-id>`
I guess desktop can generate such a link, I'm not sure. status-go already supports generating links with encoded data

Furthermore, for status-mobile, we ignore the encoded data and only use the user-id to show contact profile

> How do this links suppose to open specific channel if there is no channel id inside but only community id?

mobile just ignore this part`Ow==` and only read the community-id after `#` to open community overview, not community channel

ISSUE 5 is somehow caused by this
because for community channel, there are 
- `/cc/<encoded_channel_data>#<community-id>` status-mobile ignore encoded data, jump to community overview
- `/cc/<channel_id>#<community-id>` status-mobile read and use the channel id to jump to channel (community channel PN use this)

ISSUE 5 is caused by my code failing to distinguish these two, and it's fixed 

@pavloburykh

**Comment by pavloburykh** [comment]
> mobile just ignore this part`Ow==` and only read the community-id after `#` to open community overview, not community channel

@yqrashawn just want to understand what is the difference between 

https://status.app/cc/Ow==#zQ3shYSHp7GoiXaauJMnDcjwU2yNjdzpXLosAWapPS4CFxc11
status-app://cc/Ow==#zQ3shYSHp7GoiXaauJMnDcjwU2yNjdzpXLosAWapPS4CFxc11

AND

https://status.app/c/Ow==#zQ3shYSHp7GoiXaauJMnDcjwU2yNjdzpXLosAWapPS4CFxc11
status-app://c/Ow==#zQ3shYSHp7GoiXaauJMnDcjwU2yNjdzpXLosAWapPS4CFxc11

Why do we need those https://status.app/cc/ links if they do not open channel but community?

In your file those links are listed within community channel links section
Status URL Test 2023-11-01 14-15-09

**Comment by yqrashawn** [comment]
So basically, what happens are
1. mobile need to support the new spec at https://github.com/vacp2p/rfc/pull/602/files#diff-497acddee81a09a3aa6fb5a2aa0d0585a3ac9a9628ae0e9d4855b77309a5e7cbR41
2. there's a discussion and a decision made that mobile don't read things before `#` and after `c-or-cc-or-u/`, it's recorded at https://www.notion.so/Usage-of-link-data-7cf6c9e2a9554761baae9a1ba84e70c2?pvs=4
3. PN failed https://github.com/status-im/status-mobile/issues/17718. Because there's an old format `/cc/<0x-community-pubkey><community-channel-id>/<0x-community-pubkey>` and I removed support for this
4. to keep PN working, I add support for `/cc/<community-channel-id>#<serialized-community-id(starts with z, not 0x)>` and use it for PN

that's why there are both `/c-or-cc/<encoded community or community-channel data>#<serialized-community-id>` and they both jump to community overview @pavloburykh

**Comment by pavloburykh** [comment]
@yqrashawn are we leaving [ISSUE 4](https://github.com/status-im/status-mobile/pull/17721#issuecomment-1788774937) as a followup?

**Comment by yqrashawn** [comment]
@pavloburykh Yes. I can reproduce it but have some trouble debug it. I'll look into it tomorrow. Can you create an issue for it?

**Comment by pavloburykh** [comment]
> @pavloburykh Yes. I can reproduce it but have some trouble debug it. I'll look into it tomorrow. Can you create an issue for it?

Sure! I will create a followup once we merge this PR. I am waiting for e2e run and after that will move this PR to merge.

**Comment by pavloburykh** [comment]
@yqrashawn thank you for the PR and all the additional explanations. Ready for merge.

**Comment by yqrashawn** [comment]
rebase and update status-go version, waiting for ci

**Pull request merged**
